keras_nlp/utils/text_generation.py

```diff
@@ -134,40 +158,37 @@ def token_probability_fn(inputs):
     ```
 
     """
-    prompt = validate_prompt(prompt)
+    prompt = _validate_prompt(prompt)
 
     input_is_1d = prompt.shape.rank == 1
     if input_is_1d:
         prompt = prompt[tf.newaxis, :]
-    validate_token_probability_fn(token_probability_fn, prompt)
 
-    shape = tf.shape(prompt)
-    batch_size = shape[0]
-    length = shape[1]
+    _validate_token_probability_fn(token_probability_fn, prompt)
 
-    # Pad the prompt with `pad_token_id` to `max_length`.
-    padding = tf.fill((batch_size, max_length - length), pad_token_id)
-    prompt = tf.concat((prompt, padding), axis=1)
+    prompt, mask, batch_size, length = _pad_prompts(
+        prompt, max_length, pad_token_id
+    )
 
     def one_step(length, prompt):
         pred = token_probability_fn(prompt[:, :length])
         next_token = tf.cast(tf.argmax(pred, axis=-1), dtype=prompt.dtype)
+        next_token = tf.where(mask[:, length], prompt[:, length], next_token)
 
         # Append the next token to current sequence.
-        def add_token(args):
-            sequence, token = args
-            return tf.tensor_scatter_nd_update(
-                tensor=sequence, indices=[[length]], updates=[token]
-            )
-
-        prompt = tf.map_fn(
-            fn=add_token,
-            elems=(prompt, next_token),
-            fn_output_signature=tf.TensorSpec(
-                shape=(max_length), dtype=prompt.dtype
+        prompt = tf.tensor_scatter_nd_update(
+            tensor=prompt,
+            indices=tf.stack(
```

Comment:
I'm pretty sure there should be a way to rework this scatter with the `tf.range` call, so that we are scattering the entire column in at once.
https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update#slice_updates_2

Seems like that would be more efficient.